{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dask\n",
    "import dask.dataframe as dd\n",
    "from scipy import stats\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import subprocess as sp\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "import shutil\n",
    "import glob\n",
    "import gimmemotifs\n",
    "from pathlib import Path\n",
    "import qnorm\n",
    "from sklearn import datasets, linear_model\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.metrics import mean_squared_error, r2_score, average_precision_score\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def hue_regplot(data, x, y, hue, palette=None, **kwargs):\n",
    "    from matplotlib.cm import get_cmap\n",
    "    \n",
    "    regplots = []\n",
    "    \n",
    "    levels = data[hue].unique()\n",
    "    \n",
    "    if palette is None:\n",
    "        default_colors = get_cmap('tab10')\n",
    "        palette = {k: default_colors(i) for i, k in enumerate(levels)}\n",
    "    \n",
    "    for key in levels:\n",
    "        regplots.append(\n",
    "            sns.regplot(\n",
    "                x=x,\n",
    "                y=y,\n",
    "                fit_reg=False, \n",
    "                data=data[data[hue] == key],\n",
    "                color=palette[key],\n",
    "                **kwargs\n",
    "            )\n",
    "        )\n",
    "    \n",
    "    return regplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_dir = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1'\n",
    "Path(f\"{output_dir}\").mkdir(parents=True, exist_ok=True)       \n",
    "\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/Cell_types_files_v1_esccomp.csv'\n",
    "\n",
    "#load the genome info\n",
    "genome_path_size = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa.sizes\"\n",
    "genome_path = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# specifiy files for network generation intra comparison stromal and epi\n",
    "output_dir = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1'\n",
    "Path(f\"{output_dir}\").mkdir(parents=True, exist_ok=True)       \n",
    "\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/Cell_types_files_v1_intracomp.csv'\n",
    "\n",
    "#load the genome info\n",
    "genome_path_size = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa.sizes\"\n",
    "genome_path = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/hg38/hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_type</th>\n",
       "      <th>Accesibility_peakfiles</th>\n",
       "      <th>Merged_peakfiles</th>\n",
       "      <th>scATAC_BAMfiles</th>\n",
       "      <th>TPM_matrix</th>\n",
       "      <th>compare_with</th>\n",
       "      <th>count_table_files</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>epi</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>unused</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>stromal</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>stromal</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>unused</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "      <td>epi</td>\n",
       "      <td>/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  cell_type                             Accesibility_peakfiles  \\\n",
       "0       epi  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "1   stromal  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "\n",
       "  Merged_peakfiles                                    scATAC_BAMfiles  \\\n",
       "0           unused  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "1           unused  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...   \n",
       "\n",
       "                                          TPM_matrix compare_with  \\\n",
       "0  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...      stromal   \n",
       "1  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...          epi   \n",
       "\n",
       "                                   count_table_files  \n",
       "0  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...  \n",
       "1  /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "sample_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse binding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets first perform motif enrichment\n",
    "if not os.path.exists(f\"{output_dir}/scan_results_acc_v4.scanfile\"): \n",
    "    sp.check_call(f'nice -15 gimme scan '\n",
    "        f'-T -N 4 '\n",
    "        f'-g hg38 '\n",
    "        f'/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/ATAC_qquant/tmp/mergedpeaks3.bed '\n",
    "        f'> {output_dir}/scan_results_acc_v4.scanfile',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "LESC\n",
      "CE\n",
      "Cj\n",
      "LE\n",
      "CF\n",
      "CSSC\n"
     ]
    }
   ],
   "source": [
    "# lets do ananse binding for your cell populations\n",
    "i = 0\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    Path(f\"{output_dir}/{cell_id}\").mkdir(parents=True, exist_ok=True)       \n",
    "    bam_file = sample_data.iloc[index,3]\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/binding.h5\"): \n",
    "        print(f'running ananse binding for celltype {cell_id} using the scATAC bamfile {bam_file} with the ATAC merged peaks of all joined peaks')\n",
    "        sp.check_call(f'ananse binding '\n",
    "            f'-A {bam_file} '\n",
    "            f'-o {output_dir}/{cell_id}/ '\n",
    "            f'--jaccard-cutoff 0.1 '\n",
    "            f'-g hg38 '\n",
    "            f'-r /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/ATAC_qquant/tmp/mergedpeaks3.bed '\n",
    "            f'--pfmscorefile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/scan_results_acc_v4.scanfile '\n",
    "            f'-n 1 '    \n",
    "            f'2> {output_dir}/{cell_id}/ananse_binding_log.txt',shell = True)\n",
    "        \n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "epi\n",
      "running ananse binding for celltype epi using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ATAC_bam/epi/outs/subsets/epi.bam with the narrowpeakfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ATAC_bam/epi/outs/subsets/epi.narrowPeak\n",
      "stromal\n",
      "running ananse binding for celltype stromal using the scATAC bamfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ATAC_bam/stromal/outs/subsets/stromal.bam with the narrowpeakfile /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ATAC_bam/stromal/outs/subsets/stromal.narrowPeak\n"
     ]
    }
   ],
   "source": [
    "# lets do ananse binding for your cell populations now without the pfmscorefile and only the narrowpeak\n",
    "i = 0\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    Path(f\"{output_dir}/{cell_id}/narrow\").mkdir(parents=True, exist_ok=True)       \n",
    "    bam_file = sample_data.iloc[index,3]\n",
    "    peak_file = sample_data.iloc[index,1]\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/binding.h5\"): \n",
    "        print(f'running ananse binding for celltype {cell_id} using the scATAC bamfile {bam_file} with the narrowpeakfile {peak_file}')\n",
    "        sp.check_call(f'ananse binding '\n",
    "            f'-A {bam_file} '\n",
    "            f'-o {output_dir}/{cell_id}/narrow/ '\n",
    "            f'--jaccard-cutoff 0.1 '\n",
    "            f'-g hg38 '\n",
    "            f'-r {peak_file} '\n",
    "            f'-n 1 '    \n",
    "            f'2> {output_dir}/{cell_id}/narrow/ananse_binding_log_narrow.txt',shell = True)\n",
    "        \n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generating the binding.h5 file for ESC data ananse develop 0.3.0\n",
    "Path(f\"{output_dir}/ESC\").mkdir(parents=True, exist_ok=True)       \n",
    "if not os.path.exists(f\"{output_dir}/ESC/binding.h5\"): \n",
    "    sp.check_call(f'ananse binding '\n",
    "        f'-H /ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/H3K27ac/results/final_bam/hg38-GSM466732.samtools-coordinate.bam /ceph/rimlsfnwi/data/moldevbio/zhou/jsmits/Ananse_test_data/H3K27ac/results/final_bam/hg38-GSM663427.samtools-coordinate.bam '\n",
    "        f'-r /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/ESC_ANANSE/all_peaks.tsv '\n",
    "        f'-o {output_dir}/ESC '\n",
    "        f'--jaccard-cutoff 0.1 '\n",
    "        f'-g hg38 '\n",
    "        f'-n 4 '   \n",
    "        f'2> {output_dir}/ESC/ananse_binding_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse network\n",
    "\n",
    "Generate GRN's based on the binding.tsv output and RNAseq expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/LSC_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LESC\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LESC/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/LESC_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CE\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CE/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/CE_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "Cj\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/Cj/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/Cj_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "LE\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LE/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/LE_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CF\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/CF_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "CSSC\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CSSC/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/CSSC_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n"
     ]
    }
   ],
   "source": [
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    Path(f\"{output_dir}/{cell_id}\").mkdir(parents=True, exist_ok=True)  \n",
    "    binding_file = f\"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/{cell_id}/binding.h5\"\n",
    "    print(f'running ananse binding using the binding file {binding_file}')\n",
    "    file_name = sample_data.iloc[index,4]\n",
    "    print(f'together with the gene expression file {file_name}')\n",
    "    print(f'and the output file full_network_includeprom.txt')\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"): \n",
    "        sp.check_call(f'nice -15 ananse network '\n",
    "            f' {binding_file} '\n",
    "            f'-e {file_name} '\n",
    "            f'-o {output_dir}/{cell_id}/full_network_includeprom.txt '\n",
    "            f'--full-output '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_network_log.txt',shell = True)\n",
    "        \n",
    "#             f'-a /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/genome/genomehg38gimme/hg38/hg38.annotation.gtf.gz ' Check if this can be re-run with all tfs given the right name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "epi\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/epi/narrow/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/RNA_CPM/epi_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n",
      "stromal\n",
      "running ananse binding using the binding file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/binding.h5\n",
      "together with the gene expression file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/RNA_CPM/stromal_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n"
     ]
    }
   ],
   "source": [
    "# Generate networks from Narrow peak binding files\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    Path(f\"{output_dir}/{cell_id}\").mkdir(parents=True, exist_ok=True)  \n",
    "    binding_file = f\"{output_dir}/{cell_id}/narrow/binding.h5\"\n",
    "    print(f'running ananse binding using the binding file {binding_file}')\n",
    "    file_name = sample_data.iloc[index,4]\n",
    "    print(f'together with the gene expression file {file_name}')\n",
    "    print(f'and the output file full_network_includeprom.txt')\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/full_network_includeprom.txt\"): \n",
    "        sp.check_call(f'nice -15 ananse network '\n",
    "            f' {binding_file} '\n",
    "            f'-e {file_name} '\n",
    "            f'-o {output_dir}/{cell_id}/narrow/full_network_includeprom.txt '\n",
    "            f'--full-output '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/narrow/ananse_network_log.txt',shell = True)\n",
    "        \n",
    "#       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running ananse binding using the enhancer file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/binding.h5\n",
      "together with the CPM file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/ESC_cpm.tsv\n",
      "and the output file full_network_includeprom.txt\n"
     ]
    }
   ],
   "source": [
    "# generate the ESC network file from the binding.h5 file\n",
    "binding_file = f\"{output_dir}/ESC/binding.h5\"\n",
    "print(f'running ananse binding using the enhancer file {binding_file}')\n",
    "file_name = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/RNA_CPM/ESC_cpm.tsv\"\n",
    "print(f'together with the CPM file {file_name}')\n",
    "print(f'and the output file full_network_includeprom.txt')\n",
    "    \n",
    "if not os.path.exists(f\"{output_dir}/ESC/full_network_includeprom.txt\"): \n",
    "    sp.check_call(f'nice -15 ananse network '\n",
    "        f'{binding_file} '\n",
    "        f'-e {file_name} '\n",
    "        f'-o {output_dir}/ESC/full_network_includeprom.txt '\n",
    "        f'--full-output '\n",
    "        f'-g hg38 '\n",
    "        f'-n 1 '\n",
    "        f'2> {output_dir}/ESC/ananse_network_log.txt',shell = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ananse Influence\n",
    "\n",
    "Lets finally run Ananse influence to predict key TFs that differ between cell types.\n",
    "it will compare each network with the networks listed in the 'comparison' column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LESC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LESC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LESC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "Cj\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/Cj/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/Cj_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CF\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CF_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CSSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CSSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CSSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# NARROW test\n",
    "# cell population of interest against only ESC data\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/Cell_types_files_v1_esccomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_500000_0401\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'--select-after-join '\n",
    "            f'-i 500000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/narrow/ananse_influence_log_ESC_0401.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/epi/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/epi_stromal_pseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/epi/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/stromal_epi_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# Compare cell_types against other cells Narrow (intra epi-stromal comp)\n",
    "for index, cell_type in sample_data.iterrows(): \n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    comparisons = sample_data.compare_with[index]\n",
    "    network_file = f\"{output_dir}/{cell_id}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = f\"{output_dir}/{comparisons}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/narrow/{comparisons}_to_{cell_id}_influence_250000_2604\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/{comparisons}_to_{cell_id}_influence_250000_2604/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'--select-after-join '\n",
    "            f'-i 250000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/narrow/{comparisons}_to_{cell_id}_influence_250000_2604/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/narrow/ananse_influence_log_intra_2604.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/LSC_stromal_pseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LESC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/LESC_stromal_pseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/CE_stromal_pseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/Cj/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/Cj_stromal_pseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/stromal/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/LE_stromal_pseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/epi/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/CF_epi_pseudobulkpadj.tsv and the output file influence.txt\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CSSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/outs_v1/epi/narrow/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_intra_cpm/20220421/CSSC_epi_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# Compare cell_types against epi or stromal Narrow (INTER epi-stromal comp)\n",
    "# cell population of interest against only ESC data\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_22042022/Cell_types_files_v1_intercomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows(): \n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    comparisons = sample_data.compare_with[index]\n",
    "    network_file = f\"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/{cell_id}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = f\"{output_dir}/{comparisons}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/narrow/{comparisons}_to_{cell_id}_influence_250000_2604\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/{comparisons}_to_{cell_id}_influence_250000_2604/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'--select-after-join '\n",
    "            f'-i 250000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/narrow/{comparisons}_to_{cell_id}_influence_250000_2604/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/narrow/ananse_influence_log_intra_2604.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500000\n",
      "760668\n",
      "998967\n",
      "1217553\n",
      "1363376\n",
      "1514422\n",
      "1619553\n"
     ]
    }
   ],
   "source": [
    "# Generate union_file for no sel after join\n",
    "all_interactions_set = set()\n",
    "Path(f\"{output_dir}/influence/\").mkdir(parents=True, exist_ok=True) \n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions_0604_500000.txt'\n",
    "            \n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    diff_network = f'{output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence_diffnetwork.tsv'\n",
    "    rnet = pd.read_csv(\n",
    "                    diff_network,\n",
    "                    sep=\"\\t\",\n",
    "                    dtype=\"float64\",\n",
    "                    converters={\"source\": str,\n",
    "                               \"target\": str},\n",
    "                )\n",
    "    rnet['source_target'] = rnet['source'] + \"—\" + rnet['target']\n",
    "    rnet_set = set(rnet['source_target'])\n",
    "    all_interactions_set.update(rnet_set)\n",
    "    print(len(all_interactions_set))\n",
    "\n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions_0604_500000.txt'\n",
    "interaction_df = pd.DataFrame(data=list(all_interactions_set))\n",
    "interaction_df.to_csv(interaction_file, index=False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LESC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LESC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LESC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "Cj\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/Cj/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/Cj_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CF\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CF_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CSSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CSSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CSSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# NARROW peak file networks with no sel after join\n",
    "# cell population of interest against only ESC data\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/Cell_types_files_v1_esccomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/narrow/no_sel_after_join_0604/ESC_to_{cell_id}_influence_500000\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/no_sel_after_join_0604/ESC_to_{cell_id}_influence_500000/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'-i 500000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/narrow/no_sel_after_join_0604/ESC_to_{cell_id}_influence_500000/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/narrow/no_sel_after_join_0604/ananse_influence_log_ESC.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "448666\n",
      "648352\n",
      "850195\n",
      "1018480\n",
      "1122872\n",
      "1259717\n",
      "1371514\n"
     ]
    }
   ],
   "source": [
    "# Generate union_file for no sel after join\n",
    "all_interactions_set = set()\n",
    "Path(f\"{output_dir}/influence/\").mkdir(parents=True, exist_ok=True) \n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions_0604_500000_no_sel.txt'\n",
    "            \n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    diff_network = f'{output_dir}/{cell_id}/narrow/no_sel_after_join_0604/ESC_to_{cell_id}_influence_500000/ANANSE_influence_diffnetwork.tsv'\n",
    "    rnet = pd.read_csv(\n",
    "                    diff_network,\n",
    "                    sep=\"\\t\",\n",
    "                    dtype=\"float64\",\n",
    "                    converters={\"source\": str,\n",
    "                               \"target\": str},\n",
    "                )\n",
    "    rnet['source_target'] = rnet['source'] + \"—\" + rnet['target']\n",
    "    rnet_set = set(rnet['source_target'])\n",
    "    all_interactions_set.update(rnet_set)\n",
    "    print(len(all_interactions_set))\n",
    "\n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions_0604_500000_no_sel.txt'\n",
    "interaction_df = pd.DataFrame(data=list(all_interactions_set))\n",
    "interaction_df.to_csv(interaction_file, index=False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LESC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LESC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LESC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "Cj\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/Cj/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/Cj_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CF\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CF_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CSSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CSSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CSSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# NARROW peak files with less edges: 250000; perhaps less noise and still more relevant factors?\n",
    "# cell population of interest against only ESC data\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/Cell_types_files_v1_esccomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_250000_0604\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_250000_0604/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'--select-after-join '\n",
    "            f'-i 250000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_250000_0604/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/narrow/ananse_influence_log_ESC_0604_250.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LESC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LESC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LESC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "Cj\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/Cj/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/Cj_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LE/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CF\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CF_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CSSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CSSC/narrow/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CSSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# NARROW peak files with less edges: 100000; perhaps less noise and still more relevant factors seems like it based on 250.000\n",
    "# cell population of interest against only ESC data\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/Cell_types_files_v1_esccomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/narrow/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_100000_0804\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_100000_0804/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'--select-after-join '\n",
    "            f'-i 100000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/narrow/ESC_to_{cell_id}_influence_100000_0804/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/narrow/ananse_influence_log_ESC_0804_100.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LESC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LESC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LESC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CE/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "Cj\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/Cj/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/Cj_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LE/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/LE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CF\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CF_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CSSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CSSC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/CSSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# cell population of interest against only ESC data\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/Cell_types_files_v1_esccomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_0401\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'--select-after-join '\n",
    "            f'-i 500000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_influence_log_ESC_0401.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500000\n",
      "721884\n",
      "870189\n",
      "982089\n",
      "1138593\n",
      "1342377\n",
      "1514043\n"
     ]
    }
   ],
   "source": [
    "all_interactions_set = set()\n",
    "Path(f\"{output_dir}/influence/\").mkdir(parents=True, exist_ok=True) \n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions_0401_500000.txt'\n",
    "            \n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    diff_network = f'{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence_diffnetwork.tsv'\n",
    "    rnet = pd.read_csv(\n",
    "                    diff_network,\n",
    "                    sep=\"\\t\",\n",
    "                    dtype=\"float64\",\n",
    "                    converters={\"source\": str,\n",
    "                               \"target\": str},\n",
    "                )\n",
    "    rnet['source_target'] = rnet['source'] + \"—\" + rnet['target']\n",
    "    rnet_set = set(rnet['source_target'])\n",
    "    all_interactions_set.update(rnet_set)\n",
    "    print(len(all_interactions_set))\n",
    "\n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions_0401_500000.txt'\n",
    "interaction_df = pd.DataFrame(data=list(all_interactions_set))\n",
    "interaction_df.to_csv(interaction_file, index=False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test dif interactions inbetween corneal populations\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/LSC/full_network_includeprom.txt '\n",
    "            f'-s /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_31032022/outs_v1/CF/full_network_includeprom.txt '\n",
    "            f'-d /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE/RNA_DE/20220401/ '\n",
    "            f'--select-after-join '\n",
    "            f'-i 500000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_0401/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_influence_log_ESC_0401.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/LSC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/LSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LESC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/LESC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/LESC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/CE/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/CE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "Cj\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/Cj/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/Cj_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "LE\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/LE/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/LE_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CF\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/CF/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/CF_ESC_pseudobulkpadj.tsv and the output file influence.txt\n",
      "CSSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/CSSC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/CSSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# cell population of interest against only ESC data 200.000 edges diffnetwork\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/Cell_types_files_v1_esccomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/{cell_id}/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_200000_3103\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_200000_3103/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'--select-after-join '\n",
    "            f'-i 200000 '\n",
    "            f'--full-output '    \n",
    "            f'-o {output_dir}/{cell_id}/ESC_to_{cell_id}_influence_200000_3103/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_influence_log_ESC_3103.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500000\n",
      "801096\n",
      "969015\n",
      "1152289\n",
      "1253554\n",
      "1400970\n",
      "1508259\n"
     ]
    }
   ],
   "source": [
    "all_interactions_set = set()\n",
    "Path(f\"{output_dir}/influence/\").mkdir(parents=True, exist_ok=True) \n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions.txt'\n",
    "            \n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    diff_network = f'{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_3003/ANANSE_influence_diffnetwork.tsv'\n",
    "    rnet = pd.read_csv(\n",
    "                    diff_network,\n",
    "                    sep=\"\\t\",\n",
    "                    dtype=\"float64\",\n",
    "                    converters={\"source\": str,\n",
    "                               \"target\": str},\n",
    "                )\n",
    "    rnet['source_target'] = rnet['source'] + \"—\" + rnet['target']\n",
    "    rnet_set = set(rnet['source_target'])\n",
    "    all_interactions_set.update(rnet_set)\n",
    "    print(len(all_interactions_set))\n",
    "\n",
    "interaction_file = f'{output_dir}/influence/ESC_all_interactions_3003.txt'\n",
    "interaction_df = pd.DataFrame(data=list(all_interactions_set))\n",
    "interaction_df.to_csv(interaction_file, index=False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LSC\n",
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/LSC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/LSC_ESC_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'nice -15 ananse influence -t /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/LSC/full_network_includeprom.txt -s /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt -d /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/LSC_ESC_pseudobulkpadj.tsv -i 500000 --full-output -u /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/influence/ESC_all_interactions.txt -o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/LSC/ESC_to_LSC_influence_500000_2503/ANANSE_influence.tsv -n 1 2> /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/LSC/ananse_influence_log_ESC_intfile.txt' returned non-zero exit status 2.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m/scratch/julian/tmp/4MHa/ipykernel_33340/893840499.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_2503/ANANSE_influence.tsv\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m         sp.check_call(f'nice -15 ananse influence '\n\u001b[0m\u001b[1;32m     23\u001b[0m             \u001b[0;34mf'-t {network_file} '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m             \u001b[0;34mf'-s {net2} '\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/vol/mbconda/julian/envs/ananse_jupyter/lib/python3.8/subprocess.py\u001b[0m in \u001b[0;36mcheck_call\u001b[0;34m(*popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    362\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcmd\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    363\u001b[0m             \u001b[0mcmd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpopenargs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 364\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mretcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    365\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    366\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'nice -15 ananse influence -t /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/LSC/full_network_includeprom.txt -s /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt -d /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/RNA_DE/20220316/LSC_ESC_pseudobulkpadj.tsv -i 500000 --full-output -u /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/influence/ESC_all_interactions.txt -o /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/LSC/ESC_to_LSC_influence_500000_2503/ANANSE_influence.tsv -n 1 2> /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_16032022/outs_v1/LSC/ananse_influence_log_ESC_intfile.txt' returned non-zero exit status 2."
     ]
    }
   ],
   "source": [
    "# Rerun against ESC with interaction file for a more direct comparison\n",
    "\n",
    "# cell population of interest against only ESC data\n",
    "sample_data_file = '/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/Cell_types_files_v1_esccomp.csv'\n",
    "#Lets loop over all the comparisons Ananse needs to make:\n",
    "sample_data = pd.read_table(sample_data_file, \n",
    "                            sep = ',', comment = '#')\n",
    "\n",
    "for index, cell_type in sample_data.iterrows():\n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    print(cell_id)\n",
    "    network_file = f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"\n",
    "    print(f'running ananse influence using the network file {network_file}')\n",
    "    net2 = \"/ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/ESC/full_network_includeprom.txt\"\n",
    "    print(f'compared to the network file {net2}')\n",
    "    Path(f\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_2503\").mkdir(parents=True, exist_ok=True) \n",
    "    file_name = sample_data.iloc[index,6]\n",
    "    print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "    interaction_file = f'{output_dir}/influence/ESC_all_interactions.txt'\n",
    "    \n",
    "    if not os.path.exists(f\"{output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_2503/ANANSE_influence.tsv\"):\n",
    "        sp.check_call(f'nice -15 ananse influence '\n",
    "            f'-t {network_file} '\n",
    "            f'-s {net2} '\n",
    "            f'-d {file_name} '\n",
    "            f'-i 500000 '\n",
    "            f'--full-output '\n",
    "            f'-u {interaction_file} '\n",
    "            f'-o {output_dir}/{cell_id}/ESC_to_{cell_id}_influence_500000_2503/ANANSE_influence.tsv '\n",
    "            f'-n 1 '\n",
    "            f'2> {output_dir}/{cell_id}/ananse_influence_log_ESC_intfile.txt',shell = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running ananse influence using the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/LSC/full_network_includeprom.txt\n",
      "compared to the network file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/outs_v1/CF/full_network_includeprom.txt\n",
      "together with DEG file /ceph/rimlsfnwi/data/moldevbio/zhou/jarts/data/scANANSE_11032022/RNA_DE/20220311/LSC_CF_pseudobulkpadj.tsv and the output file influence.txt\n"
     ]
    }
   ],
   "source": [
    "# Compare cell_types against other cells (normalized expression)\n",
    "for index, cell_type in sample_data.iterrows(): \n",
    "    cell_id = sample_data.iloc[index,0]\n",
    "    comparisons = sample_data.compare_with[index]\n",
    "    if not cell_id.find(\"blobmn\") != -1:\n",
    "        #print(cell_id)\n",
    "        #print(comparisons)\n",
    "        network_file = f\"{output_dir}/{cell_id}/full_network_includeprom.txt\"\n",
    "        print(f'running ananse influence using the network file {network_file}')\n",
    "        net2 = f\"{output_dir}/{comparisons}/full_network_includeprom.txt\"\n",
    "        print(f'compared to the network file {net2}')\n",
    "        Path(f\"{output_dir}/{cell_id}/{cell_id}_influence_500000_1403\").mkdir(parents=True, exist_ok=True) \n",
    "        file_name = sample_data.iloc[index,6]\n",
    "        print(f'together with DEG file {file_name} and the output file influence.txt')\n",
    "\n",
    "        if not os.path.exists(f\"{output_dir}/{cell_id}/{cell_id}_influence_500000_1403/influence_diffnetwork.txt\"):\n",
    "            sp.check_call(f'nice -15 ananse influence '\n",
    "                f'-t {network_file} '\n",
    "                f'-s {net2} '\n",
    "                f'-d {file_name} '\n",
    "                f'-i 500000 '\n",
    "                f'--full-output '    \n",
    "                f'-o {output_dir}/{cell_id}/{cell_id}_influence_500000_1403/ANANSE_influence.tsv '\n",
    "                f'-n 1 '\n",
    "                f'2> {output_dir}/{cell_id}/ananse_influence_{comparisons}to{cell_id}log_500.txt',shell = True)\n",
    "                #f'--union-grn ' no union-grn anymore?"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
